src/napi/node_api_tsfn.c

```diff
@@ -0,0 +1,314 @@
+    napi_env env, napi_value func, napi_value async_resource,
+    napi_value async_resource_name, size_t max_queue_size,
+    size_t initial_thread_count, void* thread_finalize_data,
+    napi_finalize thread_finalize_cb, void* context,
+    napi_threadsafe_function_call_js call_js_cb,
+    napi_threadsafe_function* result) {
+  NAPI_TRY_ENV(env);
+  NAPI_ASSERT(result != NULL,
+              "Expect an non-null out result pointer on "
+              "napi_create_threadsafe_function.");
+
+  bool napi_async_inited = false, uv_async_inited = false,
+       uv_cond_inited = false, uv_mutex_inited = false, func_acquired = false;
+
+  napi_async_context async_context;
+  NAPI_INTERNAL_CALL(napi_async_init(env, async_resource, async_resource_name,
+                                     &async_context));
+  napi_async_inited = true;
+
+  uv_async_t* async_handle = IOTJS_ALLOC(uv_async_t);
+  iotjs_environment_t* iotjs_env = iotjs_environment_get();
+  uv_loop_t* iotjs_loop = iotjs_environment_loop(iotjs_env);
+  int uv_status = uv_async_init(iotjs_loop, async_handle, tsfn_async_callback);
+  if (uv_status != 0) {
+    goto clean;
+  }
+  uv_async_inited = true;
+
+  uv_cond_t* async_cond = IOTJS_ALLOC(uv_cond_t);
+  uv_status = uv_cond_init(async_cond);
+  if (uv_status != 0) {
+    goto clean;
+  }
+  uv_cond_inited = true;
+
+
+  uv_mutex_t* op_mutex = IOTJS_ALLOC(uv_mutex_t);
+  uv_status = uv_mutex_init(op_mutex);
+  if (uv_status != 0) {
+    goto clean;
+  }
+  uv_mutex_inited = true;
+
+  func = AS_NAPI_VALUE(jerry_acquire_value(AS_JERRY_VALUE(func)));
+  func_acquired = true;
+
+  iotjs_threadsafe_function_t* tsfn = IOTJS_ALLOC(iotjs_threadsafe_function_t);
+  tsfn->env = env;
+  tsfn->func = func;
+  tsfn->max_queue_size = max_queue_size;
+  tsfn->thread_count = initial_thread_count;
+  tsfn->thread_finalize_data = thread_finalize_data;
+  tsfn->thread_finalize_cb = thread_finalize_cb;
+  tsfn->context = context;
+  tsfn->call_js_cb = call_js_cb;
+
+  tsfn->async_context = async_context;
+  tsfn->async_handle = async_handle;
+  tsfn->async_cond = async_cond;
+  tsfn->op_mutex = op_mutex;
+
+  tsfn->invocation_head = NULL;
+  tsfn->invocation_tail = NULL;
+  tsfn->queue_size = 0;
+
+  tsfn->aborted = false;
+
+  async_handle->data = (void*)tsfn;
+
+  NAPI_ASSIGN(result, (napi_threadsafe_function)tsfn);
+  NAPI_RETURN(napi_ok);
+
+clean:
+  if (napi_async_inited) {
+    napi_async_destroy(env, async_context);
+  }
+  if (func_acquired) {
+    jerry_release_value(AS_JERRY_VALUE(func));
+  }
+  if (uv_async_inited) {
+    uv_close((uv_handle_t*)async_handle, tsfn_async_close_cb);
+  }
+  if (uv_cond_inited) {
+    uv_cond_destroy(async_cond);
+  }
+  if (uv_mutex_inited) {
+    uv_mutex_destroy(op_mutex);
+  }
+  NAPI_RETURN_WITH_MSG(napi_generic_failure,
+                       "Unexpected error on napi_create_threadsafe_function.");
+}
+
+napi_status napi_get_threadsafe_function_context(napi_threadsafe_function func,
```

Comment:
This function could be inlined?

Reply:
This function is declared in `node_api.h`. We should not change its function signature.

Reply:
I see.